test/common/common/regex_test.cc

```diff
@@ -50,9 +50,8 @@ TEST(Utility, ParseRegex) {
     envoy::type::matcher::RegexMatcher matcher;
     matcher.mutable_google_re2()->mutable_max_program_size()->set_value(1);
     matcher.set_regex("/asdf/.*");
-    EXPECT_THROW_WITH_MESSAGE(Utility::parseRegex(matcher), EnvoyException,
-                              "regex '/asdf/.*' RE2 program size of 24 > max program size of 1. "
-                              "Increase configured max program size if necessary.");
+    EXPECT_THROW_WITH_REGEX(Utility::parseRegex(matcher), EnvoyException,
```

Comment:
@junyer I have general question about this max program size - If it can change with RE2 implementation, how does people generally configure this? Would it not break users if it changes? Do you have any suggestions on how we can arrive the max program size for a regex and configure it here?

Reply:
Envoy itself enforces `max_program_size`; RE2 has no such configuration knob. Such a limit lets you reject regular expressions that "cost too much", but expecting anything about the absolute "cost" of any given regular expression is brittle unless it's something "obvious" like `> 1`. If RE2 compiled to LLVM bitcode, you wouldn't set any expectations around how many bytes of LLVM bitcode were compiled for any given regular expression, but you could still say "well, that's just too big" at some point. And so it is here with Envoy. :)

Reply:
Thank you for the explanation. Yeah, RE2 does not provide that knob, but Envoy provides that knob to users and I was n't sure on what basis should one set this value :-)
Another question, what is typically safe to set assuming to be on safer side for large regex? I have seen RE2, setting 9999 for its fuzz tests. The reason I am asking is - should we give users control to set the value for every regex or can we set some generic higher value (say 512/1024 or some thing that would satisfy most regexes that a deployment can use?

Reply:
The default of 100 seems a little low to me, but I don't really know what Envoy users do with regular expressions. It might be worth looking into bumping the limit up to (say) 10,000 and configuring the memory budget per `RE2` object, which is a more pragmatic approach. (The memory budget defaults to 8MiB.) I'm more than happy to discuss that approach if you like, but could that please not block this PR from landing? :)

Reply:
Sure. I just want to know the details. Definitely not to block this PR from landing :-). 

One final question - 
> The memory budget defaults to 8MiB

With default memory, what would be the ideal max program size that it can support per RE2 instance?

Reply:
Unfortunately, the answer is "it depends". Most of the memory budget is going to be used by the DFA state cache, generally speaking, and the only thing that RE2 can really do here is to bound that usage. Reasoning based on program size doesn't actually work here because the exponential worst case involves the equivalent of NFA states, not individual instructions. The DFA state cache is cleared when it would overflow, so any regular expression that is "too big" can still be executed albeit with varying levels of performance degradation. (And RE2 will fall back to NFA execution if it determines that performance has degenerated too much.)

Reply:
Great explanation. Thank you so much.